server/src/main/java/io/crate/expression/scalar/bitwise/BitwiseFunctions.java

```diff
@@ -0,0 +1,126 @@
+            new TypedOperator<>(BYTE, (a, b) -> (byte) (a & b)),
+            new TypedOperator<>(BitStringType.INSTANCE_ONE, (a, b) -> {
+                LENGTH_VALIDATOR.accept("AND", a, b);
+                a.bitSet().and(b.bitSet()); // Mutates 'a'
+                return a;
+            })
+        ),
+        OR(
+            new TypedOperator<>(LONG, (a, b) -> a | b),
+            new TypedOperator<>(INTEGER, (a, b) -> a | b),
+            new TypedOperator<>(SHORT, (a, b) -> (short) (a | b)),
+            new TypedOperator<>(BYTE, (a, b) -> (byte) (a | b)),
+            new TypedOperator<>(BitStringType.INSTANCE_ONE, (a, b) -> {
+                LENGTH_VALIDATOR.accept("OR", a, b);
+                a.bitSet().or(b.bitSet());
+                return a;
+            })
+        ),
+        XOR(
+            new TypedOperator<>(LONG, (a, b) -> a ^ b),
+            new TypedOperator<>(INTEGER, (a, b) -> a ^ b),
+            new TypedOperator<>(SHORT, (a, b) -> (short) (a ^ b)),
+            new TypedOperator<>(BYTE, (a, b) -> (byte) (a ^ b)),
+            new TypedOperator<>(BitStringType.INSTANCE_ONE, (a, b) -> {
+                LENGTH_VALIDATOR.accept("XOR", a, b);
+                a.bitSet().xor(b.bitSet());
+                return a;
+            })
+        );
+
+        private final List<TypedOperator> typedOperators = new ArrayList<>();
+
+        Operations(TypedOperator<Long> longOperator,
+                   TypedOperator<Integer> intOperator,
+                   TypedOperator<Short> shortOperator,
+                   TypedOperator<Byte> byteOperator,
+                   TypedOperator<BitString> bitStringOperator) {
+            typedOperators.add(longOperator);
+            typedOperators.add(intOperator);
+            typedOperators.add(shortOperator);
+            typedOperators.add(byteOperator);
+            typedOperators.add(bitStringOperator);
+        }
+
+        @Override
+        public String toString() {
+            return name().toLowerCase(Locale.ENGLISH);
+        }
+    }
+
+    public static void register(ScalarFunctionModule module) {
+        for (var operation : Operations.values()) {
+            for (TypedOperator typedOperator : operation.typedOperators) {
+                TypeSignature typeSignature = typedOperator.type.getTypeSignature();
+                module.register(
+                    Signature.scalar(
+                        operation.toString(),
+                        typeSignature,
+                        typeSignature,
+                        typeSignature
+                    ).withFeatures(Scalar.DETERMINISTIC_ONLY),
+                    (signature, boundSignature) ->
+                        new BinaryScalar<>(typedOperator.operator, signature, boundSignature, typedOperator.type)
+                );
+            }
+        }
+    }
```

Comment:
Could change this to do the registration directly with something like:

```java
    private static <T> void register(ScalarFunctionModule module,
                                     String name,
                                     DataType<T> type,
                                     BinaryOperator<T> operator) {
        TypeSignature typeSignature = type.getTypeSignature();
        Signature scalar = Signature.scalar(name, typeSignature, typeSignature, typeSignature)
            .withFeatures(Scalar.DETERMINISTIC_ONLY);
        module.register(scalar, (signature, boundSignature) -> new BinaryScalar<>(operator, signature, boundSignature, type));
    }


    public static void register(ScalarFunctionModule module) {
        register(module, "AND", LONG, (a, b) -> a & b);
        register(module, "AND", INTEGER, (a, b) -> a & b);
        // ...
    }
```

That would have the advantage that we don't keep the `typedOperators` lists in-memory forever

Reply:
Good point, created https://github.com/crate/crate/pull/13291.